### PR TITLE
crypto: Begin SubtleCrypto implementation

### PR DESCRIPTION
**Pull request by msub2**
This begins the implementation for SubtleCrypto by adding the generateKey function with support for AES-CBC to start. Opening as draft to get feedback on the DOM interface code first, willl update the WPT expectations when everything else is looking good. This currently lets `WebCryptoAPI/generateKey/failures_AES-CBC.https.any.html` fully pass.

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [ ] These changes fix #___ (GitHub issue number if applicable)

- [x] There are tests for these changes OR
- [ ] These changes do not require tests because ___

### CODE REVIEW

File: components/script/dom/subtlecrypto.rs [hunk 1]
```diff
@@ -0,0 +1,250 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+use std::rc::Rc;
+
+use aes::cipher::KeyInit;
+use aes::{Aes128, Aes192, Aes256};
```
Comment (mrobinson) [inline components/script/dom/subtlecrypto.rs:8]: I'm curious about the use of `aes`. Is it possible to use a higher-level library like `ring`?

**Review by mrobinson — CHANGES_REQUESTED** [review]
Nice work. I know this is still a draft, but I did a quick review.

File: components/script/dom/subtlecrypto.rs [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+/// The spec states to run operation steps in parallel, but DOMString does not impl Send.
+unsafe impl Send for NormalizedAlgorithm {}
+
+impl Clone for Algorithm {
+    fn clone(&self) -> Self {
+        Algorithm {
+            name: self.name.clone(),
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (jdm) [reply components/script/dom/subtlecrypto.rs:158]: This is just a missing feature of the current codegen. There are manual implementations of traits like Close be and PartialEq for various codegen'd ducts and enums.

File: components/script/dom/subtlecrypto.rs [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+#[derive(Clone)]
+pub enum NormalizedAlgorithm {
+    #[allow(dead_code)]
+    Algorithm(Algorithm),
+    AesKeyGenParams(AesKeyGenParams),
+}
+
+#[allow(unsafe_code)]
+/// The spec states to run operation steps in parallel, but DOMString does not impl Send.
+unsafe impl Send for NormalizedAlgorithm {}
```
Comment (jdm) [inline components/script/dom/subtlecrypto.rs:150]: This is one of those things that will bite us if we ever make use of the !Send properties of DOMString. I would prefer to explicitly convert the strings to a sendable type instead of using unsafe code here.

File: components/script/dom/subtlecrypto.rs [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+                Handle::Aes192(key)
+            },
+            256 => {
+                let key = Aes256::new_from_slice(&rand).map_err(|_| Error::Operation)?;
+                Handle::Aes256(key)
+            },
+            _ => return Err(Error::Operation),
+        };
+
+        Ok(CryptoKey::new_inherited(
```
Comment (jdm) [inline components/script/dom/subtlecrypto.rs:240]: I'm surprised to see a use of new_inherited here.

File: components/script/dom/subtlecrypto.rs [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+}
+
+impl SubtleCrypto {
+    /// <https://w3c.github.io/webcrypto/#aes-cbc-operations>
+    fn generate_key_aes_cbc(
+        &self,
+        usages: Vec<KeyUsage>,
+        key_gen_params: AesKeyGenParams,
+        extractable: bool,
+    ) -> Result<CryptoKey, Error> {
```
Comment (jdm) [inline components/script/dom/subtlecrypto.rs:206]: This should be `Root<CryptoKey>` as far as I can see.

File: components/script/dom/cryptokey.rs [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+            extractable,
+            algorithm,
+            usages,
+            handle,
+        }
+    }
+}
+
+impl CryptoKeyMethods for CryptoKey {
+    // https://w3c.github.io/webcrypto/#cryptokey-interface-members
```
[the remark this answers is not in the document]
Comment (msub2) [reply components/script/dom/cryptokey.rs:60]: Ah that's my bad, they were supposed to be rustdoc yeah. Will fix

File: components/script/dom/subtlecrypto.rs (hunk 1 above)
Comment (msub2) [reply components/script/dom/subtlecrypto.rs:8]: Potentially, according to some prior investigation by jdm though ring doesn't have support for AES-CBC https://github.com/servo/servo/issues/26878#issuecomment-658241844

File: components/script/dom/subtlecrypto.rs (hunk 3 above)
Comment (msub2) [reply components/script/dom/subtlecrypto.rs:150]: Noted, I'll rework this then

File: components/script/dom/subtlecrypto.rs [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+use crate::dom::globalscope::GlobalScope;
+use crate::dom::promise::Promise;
+use crate::realms::InRealm;
+use crate::script_runtime::JSContext;
+use crate::task_source::TaskSource;
+
+#[dom_struct]
+pub struct SubtleCrypto {
+    reflector_: Reflector,
+    #[ignore_malloc_size_of = "Defined in rand"]
```
[the remark this answers is not in the document]
Comment (msub2) [reply components/script/dom/subtlecrypto.rs:34]: I believe the MallocSizeOf issue here comes from the external `rand` crate that's used in Servo's `rand`

[the remark this answers is not in the document]
Comment (mrobinson) [reply components/script/dom/subtlecrypto.rs:34]: I think the right thing here is to move the  `#[ignore_malloc_size_of = "Defined in rand"]` decorator to the `rand` crate so this doesn't have to be repeated everywhere.

**Comment by msub2** [comment]
Alright, I believe I've addressed all the feedback so far. Let me know if there's anything major still outstanding.

**Comment by msub2** [comment]
Did a little bit more work to implement exportKey (that way both the failures and successes tests can run) and fixed some assumptions I made about how the aes crate works, so the CryptoKey handle now holds the actual key data we generated (the Aes structs I had before seem to be more opaque wrappers that take in the key data to facilitate encryption/decryption, but you can't retrieve the key data itself from them).

File: components/script/dom/subtlecrypto.rs [hunk 8]
```diff
@@ -0,0 +1,328 @@
...
+                let exported_key = match alg_name.as_str() {
+                    ALG_AES_CBC => subtle.export_key_aes_cbc(format, &*key),
+                    _ => Err(Error::NotSupported),
+                };
+                match exported_key {
+                    Ok(k) => {
+                        let cx = GlobalScope::get_cx();
+                        rooted!(in(*cx) let mut array_buffer_ptr = ptr::null_mut::<JSObject>());
+                        let _ = unsafe {
+                            ArrayBuffer::create(
```
Comment (jdm) [inline components/script/dom/subtlecrypto.rs:181]: Are you able to use the safe [create_buffer_source](https://doc.servo.org/script/dom/bindings/buffer_source/fn.create_buffer_source.html) instead?

File: components/servo/Cargo.toml [hunk 9]
```diff
@@ -32,6 +32,7 @@ webgl_backtrace = [
 ]
 
 [dependencies]
+aes = { workspace = true }
```
Comment (jdm) [inline components/servo/Cargo.toml:35]: Why is this added here?

**Review by jdm — APPROVED** [review]

File: components/servo/Cargo.toml (hunk 9 above)
Comment (msub2) [reply components/servo/Cargo.toml:35]: Ah my bad, must be left over from when I was still figuring things out, will remove.

File: tests/wpt/mozilla/tests/mozilla/interfaces.https.html [hunk 10]
```diff
@@ -239,6 +240,7 @@
   "StorageEvent",
   "StyleSheet",
   "StyleSheetList",
+  "SubtleCrypto",
```
Comment (jdm) [inline tests/wpt/mozilla/tests/mozilla/interfaces.https.html:243]: We should probably gate these new interfaces behind a pref until more if the functionality is present.

Comment (msub2) [reply tests/wpt/mozilla/tests/mozilla/interfaces.https.html:243]: Ah good point, I'll add that

**Review by jdm — APPROVED** [review]

**Comment by msub2** [comment]
@mrobinson I think you need to approve as well? Still says merging is blocked

**Review by mrobinson — APPROVED** [review]
